CMakeLists.txt

```diff
@@ -67,7 +67,17 @@ add_definitions ( "/D_CRT_NONSTDC_NO_WARNINGS")
 add_definitions( "/wd4503")
 endif()
 
-SET(DFHACK_BUILD_ARCH "32" CACHE STRING "Architecture to build ('32' or '64')")
+# Automatically detect architecture based on Visual Studio generator
+IF(MSVC AND NOT DEFINED DFHACK_BUILD_ARCH)
+    IF(${CMAKE_GENERATOR} MATCHES "Win64")
+        SET(DFHACK_BUILD_ARCH "64")
+    ELSE()
+        SET(DFHACK_BUILD_ARCH "32")
+    ENDIF()
+ELSE()
+    SET(DFHACK_BUILD_ARCH "32" CACHE STRING "Architecture to build ('32' or '64')")
```

Comment:
The options here (`CACHE STRING "Architecture to build ('32' or '64')"`) are fairly important - they make this option show up in CMakeCache.txt and various GUIs (ccmake, cmake-gui, etc.). Does this change break that? I would leave the original line as-is, except that would break the part that detects whether DFHACK_BUILD_ARCH is already defined. Maybe something like this would work:

```cmake
SET(DFHACK_BUILD_ARCH "" CACHE STRING "Architecture to build ('32' or '64')")
IF(MSVC AND "${DFHACK_BUILD_ARCH}" STREQUAL "")
    IF(${CMAKE_GENERATOR} MATCHES "Win64")
        SET(DFHACK_BUILD_ARCH "64")
    ELSE()
        SET(DFHACK_BUILD_ARCH "32")
    ENDIF()
ELSE()
    SET(DFHACK_BUILD_ARCH "32")
ENDIF()
```